### PR TITLE
New add to db

### PR DESCRIPTION
**Pull request by PhilPJL**
Please review and check that this doesn't break anything.
If you're happy then I'll remove old code and extra logging before merging.

The old AddToDB method is replaced using reflection to try to add all derived definition types to matching databases where possible.

### CODE REVIEW

File: SolastaModApi/BuilderHelpers/BaseDefinitionBuilder.cs [hunk 1]
```diff
@@ -58,6 +62,129 @@ protected BaseDefinitionBuilder(string name, string guid, string title_string, s
...
+                    {
+                        throw new SolastaModApiException($"Could not locate the 'HasElementByGuid' method for {dbType.FullName}.");
+                    }
+
+                    return (bool)methodInfo.Invoke(db, new object[] { guid });
+                }
+            }
+
+            // Get list of base types down to but not including BaseDefinition.  
+            IEnumerable<Type> GetBaseTypes(Type t)
```
Comment (ChrisPJohn) [inline SolastaModApi/BuilderHelpers/BaseDefinitionBuilder.cs:173]: This is cool. Reflection lets you do some neat things.

**Review by ChrisPJohn — CHANGES_REQUESTED** [review]
Question about the logic for the best way to get the typed databases.

**Comment by PhilPJL** [comment]
You said

_I see two possible solutions here_

_1. add support for specifying the db to get added to which would in this case let me say yeah the class is CustomizableFightingStyle but really this is a FightingStyleDefinition add it to that db. If we do this we should still check that the provided object does extend the provided database class._

This could be done by making this existing method public

`private static void AddToDB<TDatabase>(TDatabase definition, bool assertIfDuplicate = true) where TDatabase : BaseDefinition
`

That's making life more complicated compared to calling AddToDB() and expecting it to do the right thing.  

_2. Instead of attempting to add to one db most of the time, check if the provided object is a valid type for every db the game has and add it to all of the appropriate dbs._

Possible but requires reflection since the dbs are stored in:
```
public class DatabaseRepository
{
  // Field databases with token 040005CC
  private static Dictionary<System.Type, object> databases = new Dictionary<System.Type, object>();
```

_Option 3_ would be to fix the script that generated the contents of the static AddToDB() method.  AddToDB() would need to be regenerated for every release of Solasta.

_Option 4_ which I implemented is similar to option 2 except we instead look for databases that match all of the base types of the definition type.  Still needs reflection but is faster than searching the whole collection of databases.  This should be a robust solution which is fairly future proof, assuming TA don't change the names of the methods like _HasElement_.

File: SolastaModApi/BuilderHelpers/BaseDefinitionBuilder.cs [hunk 2]
```diff
@@ -58,6 +62,129 @@ protected BaseDefinitionBuilder(string name, string guid, string title_string, s
...
+
+            // TODO: remove logging
+            Main.Log($"AddToDB {Definition.GetType().FullName}, succeeded");
+
+            return Definition;
+
+            bool AddToDB(Type type)
+            {
+                // get database
+                var getDatabaseMethodInfoGeneric = GetDatabaseMethodInfo.MakeGenericMethod(type);
```
[the remark this answers is not in the document]
Comment (PhilPJL) [reply SolastaModApi/BuilderHelpers/BaseDefinitionBuilder.cs:106]: the issue is that `Type` and `<T>` don't mix so I had to go the reflection route.

**Review by ChrisPJohn — APPROVED** [review]

**Pull request merged**